### DISCUSSION TITLE
Support Circle shape in labelme2coco.py JSON generation

### DISCUSSION THREAD

**Discussion by stephanj** [wkentaro/labelme#1134] [Ideas / Feature requests]
The current labelme2coco.py doesn't support the circle shape.

**Comment by stephanj** [comment]
You can achieve this by adding following logic to /examples/instance_segmentation/labelme2coco.py:

```
        # ...
            if label in masks:
                masks[label] = masks[label] | mask
            else:
                masks[label] = mask

            # Support for circle shape by Stephan Janssen (sja@devoxx.com)
            if shape_type == "circle":
                
                # Create polygon shaped based on connecting lines from/to following degress
                bearing_angles = [0, 15, 30, 45, 60, 75, 90, 105, 120, 135, 150, 165, 
                180, 195, 210, 225, 240, 255, 270, 285, 300, 315, 330, 345, 360]
                            
                orig_x1 = points[0][0] 
                orig_y1 = points[0][1] 

                orig_x2 = points[1][0]
                orig_y2 = points[1][1]

                # Calculate radius of circle
                cx = (orig_x2 - orig_x1)**2
                cy = (orig_y2 - orig_y1)**2
                radius = math.sqrt(cx + cy)

                circle_polygon = []
            
                for i in range(0, len(bearing_angles) - 1):
                    ad1 = math.radians(bearing_angles[i])
                    x1 = radius * math.cos(ad1)
                    y1 = radius * math.sin(ad1)
                    circle_polygon.append( (orig_x1 + x1, orig_y1 + y1) )

                    ad2 = math.radians(bearing_angles[i+1])
                    x2 = radius * math.cos(ad2)  
                    y2 = radius * math.sin(ad2)
                    circle_polygon.append( (orig_x1 + x2, orig_y1 + y2) )

                points = circle_polygon

            else:
                points = np.asarray(points).flatten().tolist()
            
            segmentations[label].append(points)                

        for label, mask in masks.items():
        # ...
```

**Comment by pieterblok** [comment]
@stephanj thank you for this good addition!!! I would recommend to add one additional line of code at the end of the circle procedure, as this line will flatten the array in such a way that it corresponds to the visualization method of the cocoapi. Again, thank you for this addition. 

> You can achieve this by adding following logic to /examples/instance_segmentation/labelme2coco.py:
> 
> ```
>         # ...
>             if label in masks:
>                 masks[label] = masks[label] | mask
>             else:
>                 masks[label] = mask
> 
>             # Support for circle shape by Stephan Janssen (sja@devoxx.com)
>             if shape_type == "circle":
>                 
>                 # Create polygon shaped based on connecting lines from/to following degress
>                 bearing_angles = [0, 15, 30, 45, 60, 75, 90, 105, 120, 135, 150, 165, 
>                 180, 195, 210, 225, 240, 255, 270, 285, 300, 315, 330, 345, 360]
>                             
>                 orig_x1 = points[0][0] 
>                 orig_y1 = points[0][1] 
> 
>                 orig_x2 = points[1][0]
>                 orig_y2 = points[1][1]
> 
>                 # Calculate radius of circle
>                 cx = (orig_x2 - orig_x1)**2
>                 cy = (orig_y2 - orig_y1)**2
>                 radius = math.sqrt(cx + cy)
> 
>                 circle_polygon = []
>             
>                 for i in range(0, len(bearing_angles) - 1):
>                     ad1 = math.radians(bearing_angles[i])
>                     x1 = radius * math.cos(ad1)
>                     y1 = radius * math.sin(ad1)
>                     circle_polygon.append( (orig_x1 + x1, orig_y1 + y1) )
> 
>                     ad2 = math.radians(bearing_angles[i+1])
>                     x2 = radius * math.cos(ad2)  
>                     y2 = radius * math.sin(ad2)
>                     circle_polygon.append( (orig_x1 + x2, orig_y1 + y2) )
> 
>                 points = circle_polygon
>                 points = np.asarray(points).flatten().tolist()
> 
>             else:
>                 points = np.asarray(points).flatten().tolist()
>             
>             segmentations[label].append(points)                
> 
>         for label, mask in masks.items():
>         # ...
> ```

**Comment by xingkongliang** [comment]
> You can achieve this by adding following logic to /examples/instance_segmentation/labelme2coco.py:
> 
> ```
>         # ...
>             if label in masks:
>                 masks[label] = masks[label] | mask
>             else:
>                 masks[label] = mask
> 
>             # Support for circle shape by Stephan Janssen (sja@devoxx.com)
>             if shape_type == "circle":
>                 
>                 # Create polygon shaped based on connecting lines from/to following degress
>                 bearing_angles = [0, 15, 30, 45, 60, 75, 90, 105, 120, 135, 150, 165, 
>                 180, 195, 210, 225, 240, 255, 270, 285, 300, 315, 330, 345, 360]
>                             
>                 orig_x1 = points[0][0] 
>                 orig_y1 = points[0][1] 
> 
>                 orig_x2 = points[1][0]
>                 orig_y2 = points[1][1]
> 
>                 # Calculate radius of circle
>                 cx = (orig_x2 - orig_x1)**2
>                 cy = (orig_y2 - orig_y1)**2
>                 radius = math.sqrt(cx + cy)
> 
>                 circle_polygon = []
>             
>                 for i in range(0, len(bearing_angles) - 1):
>                     ad1 = math.radians(bearing_angles[i])
>                     x1 = radius * math.cos(ad1)
>                     y1 = radius * math.sin(ad1)
>                     circle_polygon.append( (orig_x1 + x1, orig_y1 + y1) )
> 
>                     ad2 = math.radians(bearing_angles[i+1])
>                     x2 = radius * math.cos(ad2)  
>                     y2 = radius * math.sin(ad2)
>                     circle_polygon.append( (orig_x1 + x2, orig_y1 + y2) )
> 
>                 points = circle_polygon
> 
>             else:
>                 points = np.asarray(points).flatten().tolist()
>             
>             segmentations[label].append(points)                
> 
>         for label, mask in masks.items():
>         # ...
> ```
```python
                ...
                points = circle_polygon
                points = np.asarray(points).flatten().tolist()
                ...
```

**Comment by fanweiya** [comment]
@stephanj Why are the segmentations points the same after use circle transformation? Can You upload the full script？this is my [datasets](https://cowtransfer.com/s/b2336ba1ac1642) 
![image](https://user-images.githubusercontent.com/11371707/100588331-b26d0980-332c-11eb-93b8-721dd0f61615.png)
